server/parsec/components/postgresql/invite.py

```diff
@@ -1097,55 +1374,563 @@ async def test_dump_all_invitations(
+
+    async def _claimer_step(
+        self,
+        conn: AsyncpgConnection,
+        greeting_attempt_internal_id: int,
+        step: int,
+        claimer_data: bytes,
+    ) -> bytes | StepOutcome:
+        # Check if the step is too advanced
+        if step > 0:
+            too_advanced = await conn.fetchval(
+                *_q_step_check_too_advanced(
+                    greeting_attempt_internal_id=greeting_attempt_internal_id,
+                    step=step,
+                )
+            )
+            assert too_advanced is not None
+            if too_advanced:
+                return self.StepOutcome.TOO_ADVANCED
+        # Check if the claimer data is a mismatch
+        mismatch = await conn.fetchval(
+            *_q_claimer_step_check_mismatch(
+                greeting_attempt_internal_id=greeting_attempt_internal_id,
+                step=step,
+                claimer_data=claimer_data,
+            )
+        )
+        assert mismatch is not None
+        if mismatch:
+            return self.StepOutcome.MISMATCH
+        # Update the step
+        greeter_data = await conn.fetchval(
+            *_q_claimer_step(
+                greeting_attempt_internal_id=greeting_attempt_internal_id,
+                step=step,
+                claimer_data=claimer_data,
+            )
+        )
+        if greeter_data is None:
+            return self.StepOutcome.NOT_READY
+        return greeter_data
+
+    # Transactions
+
     @override
+    @transaction
     async def greeter_start_greeting_attempt(
         self,
+        conn: AsyncpgConnection,
         now: DateTime,
         organization_id: OrganizationID,
         author: DeviceID,
         greeter: UserID,
         token: InvitationToken,
     ) -> GreetingAttemptID | InviteGreeterStartGreetingAttemptBadOutcome:
-        raise NotImplementedError
+        match await self.organization._get(conn, organization_id):
+            case Organization() as org:
+                pass
+            case OrganizationGetBadOutcome.ORGANIZATION_NOT_FOUND:
+                return InviteGreeterStartGreetingAttemptBadOutcome.ORGANIZATION_NOT_FOUND
+        if org.is_expired:
+            return InviteGreeterStartGreetingAttemptBadOutcome.ORGANIZATION_EXPIRED
+
+        match await self.user._check_device(conn, organization_id, author):
+            case (greeter_user_id, greeter_profile, _):
+                assert greeter == greeter_user_id
+                pass
+            case CheckDeviceBadOutcome.DEVICE_NOT_FOUND:
+                return InviteGreeterStartGreetingAttemptBadOutcome.AUTHOR_NOT_FOUND
+            case CheckDeviceBadOutcome.USER_NOT_FOUND:
+                return InviteGreeterStartGreetingAttemptBadOutcome.AUTHOR_NOT_FOUND
+            case CheckDeviceBadOutcome.USER_REVOKED:
+                return InviteGreeterStartGreetingAttemptBadOutcome.AUTHOR_REVOKED
+
+        invitation_info = await self.get_invitation_info(conn, organization_id, token)
```

Comment:
There is no lock here, how is handled the concurrency ? (i.e. multiple calls to `greeter_start_greeting_attempt` leading to more than one active attempt)

Reply:
This cannot happen due to :
```sql
-- Makes sure that there is only one active greeting attempt per session
CREATE UNIQUE INDEX unique_active_attempt ON greeting_attempt (greeting_session)
WHERE cancelled_on IS NULL;
```